### ISSUE TITLE
Types containing only cyclic values are treated as empty

### ISSUE THREAD

**Issue by rdhananjaya** [ballerina-platform/nballerina#1041]
**Description:**

```ballerina
type R1 record {| R2 f; |};

type R2 record {| R1 f; |};

// @type R1 < L2
type L2 string[];

// @type R1 < I
type I int;

type RL1 [int, RL2];
// @type RL1 < I
type RL2 [int, RL1];

```

Subtyping relation between strictly recursive types seems to work as expected.

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

**Comment by lochana-chathura** [comment]
Simple version: 
```bal
type L [L];

// @type N < L
type N never;
```

or with records

```ballerina
type R record {| R f; |};
// @type N < R
type N never;
```

**Comment by jclark** [comment]
Note that CDuce will treat R defined as

```
type R = { f = R }
```

as empty, but R defined as

```
type R = { f = ref(R) }
```

as non-empty (because in CDuce `ref t` desugars to `{ get = [] -> t ; set = t -> [] }`).

So this is something that our cell implementation needs to handle. But it currently doesn't: https://github.com/ballerina-platform/nballerina/issues/1048#issuecomment-1206112316.

**Comment by jclark** [comment]
Note that we can create such a value:

```ballerina
type Linked record {| any link; |};
type Cyclic record {| Cyclic link; |};

function createCyclic() returns Cyclic {
   Linked head = { link: () }; // non cyclic
   head.link = head; // now cyclic
   head = head.cloneReadOnly();  // now a readonly copy
   return <Cyclic>head; // should succeed: in fact, value belongs to readonly & Cyclic
}
```

**Comment by jclark** [comment]
Following paper (via Giuseppe) is relevant:

https://person.dibris.unige.it/ancona-davide/papers/AnconaCorradiOOPSLA2016extended.pdf

**Comment by jclark** [comment]
I don't think we should switch to co-inductive types, so for the moment I think the right solution is to change the spec to disallow this https://github.com/ballerina-platform/ballerina-spec/issues/1152

**Comment by jclark** [comment]
For nBallerina, we should then give an error for any record/map/tuple/array type descriptor that contains only infinite shapes.

We would need to do this in the front-end. We need to give an error if isEmpty is true for any such type descriptor. However, for recursive types we cannot do this at the time we resolve the type. Instead we would need to keep (in ModuleSymbols) a list of type descriptors to be checked (SemType + startPos/endPost) and then check those when all types are resolved. We can use Env.isReady to tell whether we can check the type at resolution time or later.

**Comment by jclark** [comment]
We can fix #1061 at the same time.
